text/3417-dropck-eyepatch-v3.md

```diff
@@ -0,0 +1,250 @@
+Dropping a type containing a lifetime which is no longer live is accepted if that lifetime is never accessed:
+```rust
+struct MyType<'s> {
+    reference: &'s str,
+    needs_drop: String,
+}
+fn can_drop_dead_reference() {
+    let _x;
+    {
+        let temp = String::from("I am only temporary");
+        _x = MyType {
+            reference: &temp,
+            needs_drop: String::from("I have to get dropped"),
+        };
+    }
+    // We drop `_x` here even though `reference` is no longer live.
+    //
+    // This is fine as dropping a reference is a noop and does not
+    // acess the pointee.
+}
+```
+The above example will however fail if we add a manual `Drop` impl as the compiler conservatively
+assumes that all generic parameters of the `Drop` impl are used:
+[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e604bcaecb7b2b4cf7fd0440faf165ac).
+
+In case a manual `Drop` impl does not access a generic parameter, you can add
+`#[may_dangle]` to that parameter. This **unsafely** asserts
+that the parameter is either completely unused when dropping your type or only
+recursively dropped. For type parameters, you have to declare whether you
+recursively drop instances of `T`. If so, you should use `#[may_dangle(droppable)]`.
+If not, you may use `#[may_dangle(must_not_use)]`.
+
+```rust
+struct MyType<T> {
+    generic: T,
+    needs_drop: String,
+}
+// The impl has to be `unsafe` as the compiler may not check
+// that `T` is actually unused.
+unsafe impl<#[may_dangle(droppable)] T> Drop for MyType<T> {
+    fn drop(&mut self) {
+        // println!("{}", self.generic); // this would be unsound
+        println!("{}", self.needs_drop);
+    }
+}
+fn can_drop_dead_reference() {
+    let _x;
+    {
+        let temp = String::from("I am only temporary");
+        _x = MyType {
+            generic: &temp,
+            needs_drop: String::from("I have to get dropped"),
+        };
+    }
+    // We drop `_x` here even though `reference` is no longer live.
+    //
+    // This is accepted as `T` is marked as `#[may_dangle(droppable)]` in the
+    // `Drop` impl of `MyType`.
+}
+```
+
+`Drop` impls for collections tend to require `#[may_dangle(droppable)]`:
+
+```rust
+pub struct BTreeMap<K, V> {
+    root: Option<Root<K, V>>,
+    length: usize,
+}
+
+unsafe impl<#[may_dangle(droppable)] K, <#[may_dangle(droppable)] V> Drop for BTreeMap<K, V> {
+    fn drop(&mut self) {
+        // Recursively drops the key-value pairs but doesn't otherwise
+        // inspect them, so we can use `#[only_dropped]` here.
+        drop(unsafe {ptr::read(self) }.into_iter())
+    }
+}
+```
+
+A type where `#[may_dangle(must_not_use)]` would be useful is a `Weak` pointer for a variant of `Rc`
+where the value is dropped when the last `Rc` goes out of scope. Dropping a `Weak` pointer
+would never access `T` in this case.
```

Comment:
This is helpful and avoids the main problem I saw in the previous RFC -- it'd be great to expand a bit on where the two modes should be used in the stdlib to help reader understand the role of them. Also, what is an example of code that compiles because of this but wouldn't otherwise?